Navigate up and down.

sriramgd opened the pull request:
For https://github.com/OmniSharp/omnisharp-roslyn/issues/97

sriramgd commented:
This is ready for review.

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        [HttpPost("navigateup")]
+        public async Task<IActionResult> NavigateUp(Request request)
```
put the Controller actions above the private methods.... it's a convention we have

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+        private async Task<IActionResult> Navigate(Request request, bool toBreak, Func<Request, FileMemberElement, GotoDefinitionResponse, bool> navigateFunc)
+        {
+            var stack = new List<FileMemberElement>(await StructureComputer.Compute(_workspace.GetDocuments(request.FileName)));
+            var ret = new GotoDefinitionResponse();
```
Shouldn't really be returning a GotoDefinitionResponse here (although we need an object with the same signature)

nosami replied:
ret is a bad name.

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+                }
+                stack.AddRange(node.ChildNodes);
+            }
+            return new ObjectResult(ret);
```
Just return the object instead of the ObjectResult. It will make it simpler to test and we need POCOs for the stdout interface I believe.

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+    {
+        private Func<Request, FileMemberElement, GotoDefinitionResponse, bool> NavigateUpFunc = (request, node, qf) =>
+        {
+            bool retVal = false;
```
retVal is a bad name

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+            bool retVal = false;
+            if (node.Location.Line < request.Line || (node.Location.Line == request.Line && node.Location.Column < request.Column))
+            {
+                qf.Line = node.Location.Line;
```
what's qf?

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+            bool toBreak = true;
+            return await Navigate(request, toBreak, NavigateDownFunc);
+        }
+
```
delete this line

nosami commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,74 @@
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc;
+using OmniSharp.Models;
+using System;
```
`using System` should be first

sriramgd commented:
Updated based on code review. Refactored not to depend on any implicit ordering of nodes. Now both navigate up and down are symmetric and work the same way.
Tested with properties, methods and classes.

davidfowl commented:
Where are the unit tests?

davidfowl commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,75 @@
+            return response;
+        }
+
+        private Func<FileMemberElement, FileMemberElement, Request, bool> IsCloserNodeUpFunc = (candidateClosestNode, closestNode, request) =>
```
Why re these funcs instead of methods?

sriramgd replied:
Added unit tests - NavigateUpDownFacts.cs.
I could use methods, but I would then need either a delegate definition to pass the methods in as parameters to the Navigate method or an if-else statement in the Navigate method.

davidfowl replied:
You can pass a method group. You don't have to define them as Func<...> to pass them as functions. The compiler does that for you. Turn these into methods.

sriramgd replied:
Done. Didnt know that before, thanks.

jrieken commented on src/OmniSharp/Models/NavigateResponse.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public class NavigateResponse
+    {
+        public string FileName { get; set; }
```
will you ever end up in a different file? Or should this be a a _code pointer_ thing (line and position)? Iff so it should have a different name

sriramgd replied:
This will not end up in a different file. Does not try follow a partial, since there is no real order for the partial. Is there any use case you had in mind?

jrieken replied:
Maybe not have the FileName than?

sriramgd replied:
Good point, fixed.

jrieken commented on src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs:
```diff
@@ -0,0 +1,75 @@
+
+            //If there is a closest node, use its line and column.
+            //or if we are on the last node, adjust column.
+            //if we are above the first or below the first node, do nothing.
```
how about (optionally, defined by request) starting at the other end and you are the top/bottom most node?

sriramgd replied:
Since this will be additive and not a breaking update, could we do this as a separate PR?

Yantrio commented:
Do we want this for all filemembers, or just methods/properties ?

sriramgd commented:
Updated based on second review. Unit tests added. Ready for another review.

davidfowl commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,845 @@
+
+namespace OmniSharp.Tests
+{
+    enum NavigateDirection {UP, DOWN}
```
This looks weird. Can you expand this so that it looks like:

```
enum NavigateDirection 
{
    UP, 
    DOWN
}
```

sriramgd replied:
Done.

nosami commented:
Nitpick - these tests would be clearer if we had another symbol in the source (similar to $) to show where the cursor will finish up.

nosami commented:
Also, am I correct in thinking that the cursor finishes up on the method name, not the first character of the line? Is that what we want?

nosami commented:
I added some keybindings for this over at https://github.com/nosami/omnisharp-demo/commit/9f374dbd1658e63d6071fb3d407a927fa18f1a72

Works well!

sriramgd commented:
Tests refactored to use a source and target marker. I prefer ending up on the definition of the class/method/property/field. Maybe the client can move to start of line etc. as required? Comments squashed in git (I think).

Ready for third review.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+{
+    enum NavigateDirection 
+    {
+      UP,
```
Indentation is off.

sriramgd replied:
Done.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using OmniSharp.Filters;
```
System.Collections.Generic, System.Linq and OmniSharp.Filters aren't used.

sriramgd replied:
Done, sorry missed that.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+    public class NavigateUpDownFacts
+    {
+
+        private string foo = "test";
```
what's this? :) delete the line above and below too

sriramgd replied:
Was testing property navigation using this file :) Removed.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+            }
+        }
+
+        private async Task AssertPosition(string fileContent, NavigateDirection navigateDirection, int finalCursorColumnOffset = 1) 
```
I don't see why this is needed. 

Should always be 0 or some constant. It shouldn't ever change. Looks as if it's masking some bug in the test code, because I can see that this code works.

sriramgd replied:
Right. Now I realize I went adjusting for a bug instead of fixing the cause. Fixed.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+
+        private string foo = "test";
+
+        private async Task<NavigateResponse> SendRequest(OmnisharpWorkspace workspace, string fileName, string fileContent, NavigateDirection upOrDown)
```
not critical, but I'd prefer it if test helper methods came after the tests.

sriramgd replied:
Done.

nosami commented on tests/OmniSharp.Tests/NavigateUpDownFacts.cs:
```diff
@@ -0,0 +1,751 @@
+                {
+                    private string text;
+
+                    public string M%oreText 
```
Why are some of the these percent markers one character in?

sriramgd replied:
Was a bug in the test code. Fixed.

sriramgd commented:
Ready for fourth :) review.

nosami commented:
Looks good to me. Thanks for this :)
